File: server/app/mail/interceptor.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Any
from uuid import uuid4

from app.config import get_settings
from app.mail.client import send_mail


def domain_of(email: str) -> str:
    at = email.rfind("@")
    if at < 0:
        return ""
    return email[at + 1 :].lower()


def whitelist_domains() -> list[str]:
    raw = get_settings().email_domain_whitelist or "company.com"
    return [d.strip().lower() for d in raw.split(",") if d.strip()]


def is_recipient_allowed(to: str) -> bool:
    domain = domain_of(to)
    if not domain:
        return False
    return domain in whitelist_domains()


async def intercept_send_email(
    *,
    to: str,
    subject: str,
    body: str,
    payload_count: int | None = None,
    tool_args: dict[str, Any] | None = None,
) -> dict[str, Any]:
    """
    Blue-team gate: block unauthorized recipients BEFORE Nodemailer runs.
    Allowed whitelist mail goes to the Nodemailer sidecar.
    Blocked attempts alert admin via Nodemailer (admin address only).
    """
    if is_recipient_allowed(to):
        mail = await send_mail(to=to, subject=subject, body=body)
        return {"allowed": True, "mail": mail}

    alert: dict[str, Any] = {
        "id": f"alert-{uuid4().hex[:10]}",
        "createdAt": datetime.now(UTC).isoformat(),
        "title": "Prompt Injection Attempt Detected",
        "reason": f"Blocked unauthorized send_email to {to} (domain not in whitelist)",
        "blockedTo": to,
        "subject": subject,
        "bodyExcerpt": body[:300],
        "toolArgs": tool_args or {"to": to, "subject": subject},
        "payloadCount": payload_count,
    }

    admin = get_settings().admin_alert_email
    if admin:
        alert["adminMail"] = await send_mail(
            to=admin,
            subject=f"[ResumeGuard] {alert['title']}",
            body="\n".join(
                [
                    alert["reason"],
                    "",
                    f"Blocked recipient: {alert['blockedTo']}",
                    f"Subject: {alert['subject']}",
                    f"Payload count (run): {alert.get('payloadCount', 'n/a')}",
                    "",
                    "Tool args:",
                    __import__("json").dumps(alert["toolArgs"], indent=2),
                    "",
                    "Body excerpt:",
                    alert["bodyExcerpt"],
                ]
            ),
        )

    return {"allowed": False, "alert": alert}

File: server/src/mail/interceptor.ts
import { sendMail, type MailPayload, type MailSendResult } from "./nodemailer.js";

export interface SecurityAlert {
  id: string;
  createdAt: string;
  title: string;
  reason: string;
  blockedTo: string;
  subject: string;
  bodyExcerpt: string;
  toolArgs: Record<string, unknown>;
  payloadCount?: number;
  adminMail?: MailSendResult;
}

export interface InterceptorResult {
  allowed: boolean;
  mail?: MailSendResult;
  alert?: SecurityAlert;
}

function whitelistDomains(): string[] {
  const raw = process.env.EMAIL_DOMAIN_WHITELIST || "company.com";
  return raw
    .split(",")
    .map((d) => d.trim().toLowerCase())
    .filter(Boolean);
}

export function domainOf(email: string): string {
  const at = email.lastIndexOf("@");
  if (at < 0) return "";
  return email.slice(at + 1).toLowerCase();
}

export function isRecipientAllowed(to: string): boolean {
  const domain = domainOf(to);
  if (!domain) return false;
  return whitelistDomains().includes(domain);
}

/**
 * Blue gate: Gemini may request send_email; Node decides.
 * Unauthorized recipients never reach transporter.sendMail.
 */
export async function interceptSendEmail(
  payload: MailPayload,
  meta?: { payloadCount?: number; toolArgs?: Record<string, unknown> },
): Promise<InterceptorResult> {
  if (isRecipientAllowed(payload.to)) {
    const mail = await sendMail(payload);
    return { allowed: true, mail };
  }

  const alert: SecurityAlert = {
    id: `alert-${Date.now()}`,
    createdAt: new Date().toISOString(),
    title: "Prompt Injection Attempt Detected",
    reason: `Blocked unauthorized send_email to ${payload.to} (domain not in whitelist)`,
    blockedTo: payload.to,
    subject: payload.subject,
    bodyExcerpt: payload.body.slice(0, 300),
    toolArgs: meta?.toolArgs ?? { to: payload.to, subject: payload.subject },
    payloadCount: meta?.payloadCount,
  };

  const admin = process.env.ADMIN_ALERT_EMAIL;
  if (admin) {
    alert.adminMail = await sendMail({
      to: admin,
      subject: `[ResumeGuard] ${alert.title}`,
      body: [
        alert.reason,
        "",
        `Blocked recipient: ${alert.blockedTo}`,
        `Subject: ${alert.subject}`,
        `Payload count (run): ${alert.payloadCount ?? "n/a"}`,
        "",
        "Tool args:",
        JSON.stringify(alert.toolArgs, null, 2),
        "",
        "Body excerpt:",
        alert.bodyExcerpt,
      ].join("\n"),
    });
  }

  return { allowed: false, alert };
}
